[WIP] Layout: Explore adding in sticky position support within the layout support

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

🚧 🚧 🚧 🚧 🚧 WIP: This is nowhere near ready for testing yet 🚧 🚧 🚧 🚧 🚧 

This PR is an early exploration into storing `position` and related data as part of the Layout object, in an effort to come up with a solution for #30121. This expands on ideas earlier explored in #38039.

Related to: #30121

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It'll be useful for many kinds of sites to be able to set one or two group blocks or template areas to be `sticky` positioned in order to anchor a header or footer to the top or the bottom of the page.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

TBC

Note: much of this is hacked together for now — all the changes in 6.1 files should be moved elsewhere prior to stabilising the PR.

## To-do

* [x] Wire up the `attributes.layout.position` value to style output in both the editor and on the site frontend (e.g. in the layout output in `layout.php`)
* [x] Switch `layout.position` to be an object, as there are likely related values that need to be stored. For example, we need the position type (e.g. `sticky`) as well as the particular edge it should be close to `top`, `bottom`, etc. Also, we might want an offset value if elements are to be set slightly off from being adjacent to the edge of the viewport.
* [x] Include related rules to offset based on the presence of the logged in admin area.
* [ ] In the output include an inferred `z-index` value to ensure that the sticky positioned element sits where it should. Edit: at the moment this is hard-coded.
* [ ] In the editor, figure out how we deal with `z-index` and ensuring that the element can be reached appropriately via List View, clicking on the element, etc, and all inspector / toolbar controls are still behaving as expected
* [ ] Even with a hard-coded z-index value, the block toolbar appears to scroll off the screen when using sticky position
* [ ] Might not be required for an MVP, but me might also want to look into overriding `scroll-padding-top` when setting a header in the site editor (e.g. [this rule](https://github.com/WordPress/wordpress-develop/blob/29b78bd67956f0ef0d7399b81427671164bc5312/src/wp-content/themes/twentysixteen/style.css#L2779))
* [ ] Explore the right place to store the `position` object in block attributes. Should it live in the `layout` object, or in the `style` object? There are trade-offs either way.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

***For testing purposes, for now, let's focus on looking just at the Group block***

1. Open up the block editor, add a group block and put a paragraph inside it, and give the group block a background color
2. In the inspector controls, set the Area to top/header, and position to Sticky
3. Scroll down, and the block should stick to the top of the screen
4. Publish the post or page, and on the site frontend, the block should stick to the top as you scroll, and factor in the height of the logged-in admin bar when it is visible

## Screenshots or screencast <!-- if applicable -->

WIPs:

| Editor | Site frontend|
| --- | --- |
| <img src="https://user-images.githubusercontent.com/14988353/195264634-141357c0-e482-43cf-94f2-3d4901c73561.gif" /> | <img src="https://user-images.githubusercontent.com/14988353/195512494-dc6d1b22-e48a-4b6a-99f7-c012bb2fc0f2.gif" /> |

andrewserong committed:
Layout: Explore adding in position support within the layout object

andrewserong committed:
Add side to data

andrewserong committed:
Add position CSS output to layout in editor view

andrewserong commented:
I've experimented in https://github.com/WordPress/gutenberg/pull/44723/commits/b4d0801539a1de1b3b06c5312961cf6bc2b5ce9e with adding position output to the layout support, so that the rules are output as part of those rules (and so we can add additional rules as needed). It seems to be working okay in the editor for now (with a hard-coded magic value for `z-index` so that it works correctly):

![2022-10-12 17 12 58](https://user-images.githubusercontent.com/14988353/195264634-141357c0-e482-43cf-94f2-3d4901c73561.gif)

Will continue hacking on this tomorrow and have a go at adding the server-side output, too.

andrewserong commented:
Another challenge, for both z-index and popovers, is figuring out how to ensure that the toolbar doesn't scroll off the page when a sticky block is selected: 🤔

![2022-10-13 10 09 42](https://user-images.githubusercontent.com/14988353/195464283-2f08a96f-cbb1-4591-a427-9c460134d01c.gif)

andrewserong committed:
Add in server-rendered output

andrewserong commented:
In https://github.com/WordPress/gutenberg/pull/44723/commits/27b2a435b4fe32e1859e241816496d7d5f90cc57, I've added in server-rendering support, which also factors in the admin bar for the top position when using a fixed or sticky position type:

![2022-10-13 16 47 11](https://user-images.githubusercontent.com/14988353/195512494-dc6d1b22-e48a-4b6a-99f7-c012bb2fc0f2.gif)

A couple of further thoughts from that change:

* For fixed headers, I imagine we'll likely want to eventually add a `scroll-padding-top` offset (like [this rule](https://github.com/WordPress/wordpress-develop/blob/29b78bd67956f0ef0d7399b81427671164bc5312/src/wp-content/themes/twentysixteen/style.css#L2779)), but it may not be a blocker for an MVP
* I'm pleased that server-rendering appears to be working pretty well, but I'm unsure if the `layout` object is the right place for these values to live — server-rendering seems to be the right way to go, but perhaps `position` should live in `style` so that isn't _quite_ so coupled to layout. In terms of block attributes, maybe `style.layout.position`? E.g. if in the future the layout classes move to the innerBlock wrapper, then the position should still be applied to the _outer_ wrapper — that seems to indicate there'll be at least some difference there
* Still need to figure out a good default `z-index` value that plays nicely in the editor (e.g. with the block toolbar), as well as on the site frontend
* How much should occur in the style engine versus in the block support (the perennial question)
* This feature requires adding additional position-related CSS properties to the list of allowed CSS properties — this may not be desired. Along with the existing cases for `display`, we'll either want to pursue whether or not to add these properties as allowed in core, or if we want an escape hatch in the style engine for particular allow-listed CSS properties — ones that we allow in certain circumstances, but not globally. Just a thought.

andrewserong committed:
Fix phpcs issue

andrewserong commented:
I'm wrapping up for the week, and busy with other things next week, so will pick this up again after that. Just jotting down some additional notes, after brainstorming a little with @tellthemachines and @ramonjd.

Tasks to try next:

* Flatten data structure for `position` object so that the props resemble the CSS properties more closely, move to `style.layout.position` possibly, so that either in this PR or in the future, we can hook in the style engine for processing the position from the style object if need be.
* Add `position`’s own output / attachment to wrapper element so that the default layout approach can safely move to innerBlocks and not worry about what position is doing. This means that the work in https://github.com/WordPress/gutenberg/pull/44600 would not have to be concerned with the needs of the position output (which should always be attached to the outer wrapper).
* Try out the lowest reasonable `z-index` value — `z-index` appears to be required because in the block editor, blocks are set to `relative` via the `.block-editor-block-list__block` classname (most likely so that toolbars, etc can be neatly overlaid). It'd be good to try to settle on a `z-index` value that just works, and allows things like the navigation block menu popover to be overlaid.

javierarce commented:
Testing this PR I noticed that changing the "Inner blocks use content width" setting resets both the Area and Position settings.

andrewserong commented:
> Testing this PR I noticed that changing the "Inner blocks use content width" setting resets both the Area and Position settings.

Thanks! I'll fix that up.

tellthemachines commented on lib/block-supports/layout.php:
```diff
@@ -25,6 +25,64 @@ function gutenberg_register_layout_support( $block_type ) {
+				'declarations' => array(
+					'position'     => $position_type,
+					$position_side => $offset_value,
+					'z-index'      => '250', // TODO: This hard-coded value should live somewhere else.
```
Can we get away with a lower value? In testing it seems to work fine with `2`. In the editor the current value is interfering with the toolbars and block appenders so in any case will need a bit of tweaking!

andrewserong replied:
Yes, we can definitely play with the value here. I'll have a play with a lower value 👍

tellthemachines commented on packages/block-editor/src/hooks/layout.js:
```diff
@@ -132,7 +133,8 @@ export function useLayoutStyles( block = {}, selector ) {
 
-function LayoutPanel( { setAttributes, attributes, name: blockName } ) {
+function LayoutPanel( props ) {
+	const { setAttributes, attributes, name: blockName } = props;
```
Was this an intended change?

andrewserong replied:
Yes, it was so that `props` can be passed down to `PositionEdit` (so that `PositionEdit` can use a similar structure as `PaddingEdit`, etc).

tellthemachines replied:
Oooh got it! Somehow missed the use of `props` below 🤦

tellthemachines commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -0,0 +1,169 @@
+					isBlock
+				>
+					{ POSITION_OPTIONS.map( ( option ) => (
+						<ToggleGroupControlOption
```
The interaction here feels a bit awkward, I wonder if it could just be a toggle to "stick block to top or bottom" which when toggled on would show the area controls?

andrewserong replied:
Oh, good point! Yes, happy to play with the design. In this PR I was mostly copying what's in the screenshots in #30121, but it's pretty bulky having both an Area and a Position section. It'd be good if we can come up with a way to group them together. I like your idea of a toggle... I wonder how it'd work if and when we have `fixed` as an option in addition to `sticky`?

tellthemachines reviewed (commented):
Thanks for this most interesting experiment! 

My first attempt at testing didn't work, and it took me a while to work out that the sticky position only works if the container of the sticky block is bigger than the sticky block itself. This can become a major usability obstacle because the block nesting structure has to be just so for it to work as expected. I'm not sure what the best way to tackle this might be, but some thoughts:

* add some guiding text when "sticky" is selected as an option, to the effect that the block will only stick to its immediate container;
* could we only enable this support for blocks that aren't the only child of their container? Or even only for top-level blocks? 
* build some custom sticky functionality with JS instead of the native CSS? 🙈 Clutching at straws here 😅

andrewserong commented:
Thanks for taking this for a spin @tellthemachines!

> I'm not sure what the best way to tackle this might be, but some thoughts:

Good thoughts, there! I think it'd be good to see how far we can get with native CSS output rather than adding in JS, and try out some of the suggested ideas about when we do or do not show the controls.

While the current use case is thinking of a global header or footer, another potential one is for headings in a glossary that stick to the top as you scroll down a really long page. Having a block that's intentionally within the post content area of a page template could be useful in that sense, so that it's not globally sticky. However allowing that kind of thing in a user friendly way is probably not simple!

Another thought is that `sticky` behaves quite differently in CSS to `fixed` in that it relates to the parent container instead of globally, so I wonder if it'll be useful for users to be able to choose between the two (though exposing `fixed` would likely create a bunch of additional challenges in the editor UI).

To expand on some of your suggestions, some other potential considerations for the controls:

* Only allow in the site editor / template editor?
* Only allow at root level, or first child of a template part?

In terms of usability, there's also the idea that if we refactor the Layout panel to use the ToolsPanel, then we might have these controls be hidden by default. In that case, folks would (hopefully) only be reaching for them when they need them, which might possibly help mitigate the usability issues a bit, though that's probably not safe to rely on 🤔

Thanks again for the thoughts and testing everybody!

andrewserong committed:
Move data to style.layout.position

andrewserong commented:
I've pushed a couple of code quality changes in https://github.com/WordPress/gutenberg/pull/44723/commits/0b7cbb245156759348794e94c551ff9ed3a9d4ab. This flattens the data a little, and moves it over to the `style` object, which is a better home for things that more closely resemble CSS values. The new structure also allows for offset values in the future (say, if someone wants their block to not be directly adjacent to the edge of the viewport). The data is now stored in `style.layout` in the block's attributes. So in real-world usage, here is a sticky group block set to the top position:

```html
<!-- wp:group {"style":{"layout":{"position":"sticky","top":"0px"}},"backgroundColor":"vivid-red","textColor":"background","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-background-color has-vivid-red-background-color has-text-color has-background"><!-- wp:paragraph -->
<p>A paragraph</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

* `style.layout.position` — a string containing the position type, e.g. `sticky`, `fixed`
* `style.layout.top` — the value for the top position
* `style.layout.right` — the value for the right position
* `style.layout.bottom` — the value for the bottom position
* `style.layout.left` — the value for the left position

This structure also allows us to add other values in the future, if we wish to set them at the block level. E.g. `style.layout.zIndex`. It also fixes the issue that @javierarce mentioned above, where toggling the 'use content width' toggle cleared out the position values. The position values should now be preserved when adjusting other layout attributes. I've also enabled the top / footer buttons to toggle the values on and off.

Still to do from [the earlier to-do list](https://github.com/WordPress/gutenberg/pull/44723#issuecomment-1278502775) (in addition to exploring the more recent feedback):

* [x] Add position’s own output / attachment to wrapper element so that the default layout approach can safely move to innerBlocks and not worry about what position is doing. (This needs to be updated in both JS and PHP)
* [ ] Try out the lowest reasonable z-index value, as in the comments above from @tellthemachines

andrewserong committed:
Fix linting issues

andrewserong committed:
Move server rendering to a separate render callback, use html tag processor to ensure output at outer wrapper level

andrewserong committed:
Move JS implementation up out of the individual layout types and give it a separate container classname

andrewserong committed:
Remove unneeded whitespace changes

andrewserong committed:
Fix up output so that position classes are applied to the wrapper

ramonjd commented on lib/block-supports/layout.php:
```diff
@@ -461,6 +461,114 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+					'sticky' === $position_type
+				) {
+					// TODO: wrap the following value in a `calc()` + `$side_value`,
+					// so that any included value is treated as an offset.
```
Will this mean that we can set `0px` as the offset for the bottom "side"?

Just thinking of the case where I'd want my footer to stick to the true bottom:

<img width="400" alt="Screen Shot 2022-10-28 at 3 49 50 pm" src="https://user-images.githubusercontent.com/6458278/198505533-a0b6a388-13e6-4983-a024-aed6569ecdd6.png">

andrewserong replied:
Oh, great catch! That particular `var()` usage with the toolbar height should only be applied on `top` — I'll fix it up 👍

andrewserong replied:
Fixed in https://github.com/WordPress/gutenberg/pull/44723/commits/5fe14685f233584f3e3dce1af7e1336b0f3636c3

ramonjd commented:
This is testing very well manually, thank you @andrewserong !!

I created a new template!

https://user-images.githubusercontent.com/6458278/198504939-feb3ef28-caaa-4843-9823-d3b39f8e0fa0.mp4

Maybe you know about this, but this is what the layout controls look like in Global styles:

<img width="278" alt="Screen Shot 2022-10-28 at 3 46 40 pm" src="https://user-images.githubusercontent.com/6458278/198504960-ee91ac83-9c31-49a1-9a98-8cf49062e4b6.png">

andrewserong commented:
Thanks for taking this for a spin @ramonjd!

> Maybe you know about this, but this is what the layout controls look like in Global styles:

This issue is currently on trunk, too, and a fix is being explored / discussed over in: https://github.com/WordPress/gutenberg/pull/45340

ramonjd commented:
> This issue is currently on trunk, too, and a fix is being explored / discussed over in: https://github.com/WordPress/gutenberg/pull/45340

Oh, my bad. I should have checked before typing my big mouth. Carry on!

andrewserong commented:
> Oh, my bad. I should have checked before typing my big mouth.

No, no, I was glad you pointed it out! I didn't realise there was a regression there, either 🙂

andrewserong committed:
Fix issue where admin height var was being used in bottom values

andrewserong committed:
Add calc for top output to support non-zero offsets that factor in the logged in admin bar height

andrewserong commented:
Update: I'll be AFK off and on over the next couple of weeks, so just jotting down some notes and thoughts for next things to work on with this PR, for when I come back to it:

* A blocker for this PR is that we will need to add position support for template parts, as for `sticky` to work correctly at the site-wide level (for a global header or footer), it needs to be applied to root level blocks. This way you could apply sticky directly to the Header or Footer template parts, instead of setting it on the first Group child of those blocks. However, if you opt-in the template part block to `__experimentalLayout`, and go to update a template part in the site editor to set the position and then hit save, it does not save the block attributes. I imagine there's some nuance here surrounding what happens that the template part block wrapper level, and what happens within a template part block? This will be something to look into.
* The block toolbar positioning doesn't play nicely with sticky positioning, in that if you scroll the window, the toolbar moves as if the block is static rather than sticky, so we lose the toolbar.
* We'll need to decide how things should behave on mobile, or across mobile and desktop. I.e. a site with a large header on desktop that can comfortably be sticky, but on mobile should be static, or a header that should be sticky on mobile but static on desktop.
* Still need to experiment with a good baseline z-index value and try out the idea of setting quite a low value.
* The controls are fairly clunky and will be infrequently used, so it might be best to explore switching the LayoutPanel to the ToolsPanel before trying to stabilise this PR (https://github.com/WordPress/gutenberg/issues/44560).
* The controls will need to include being able to select `top` or `bottom` at least, as well as normal / sticky.

jameskoster commented:
Nice work on this so far. Sticky positioning is working well for root level group blocks on the frontend, but I did notice that the effect isn't present in the editor.

> A blocker for this PR is that we will need to add position support for template parts

I'm not convinced we should do this, and would not consider it a blocker to this PR. My general feeling is that position is an abstract concept that can be applied to any block _within the context of a document_. That is to say: it's the responsibility of a _template_ to define whether its header is sticky or not.

One thing that seems important is for the values in the UI to match the CSS behaviour. IE a "Sticky" block should effectively just have `position: sticky` applied. This means it won't be possible to fix a footer to the bottom of the viewport until we add fixed positions. I think that is fine. Getting the experience for sticky positioning right is a big task on its own, and we shouldn't get bogged down in creating specific experiences for template parts yet. I think it would be fine to focus purely on adding position: sticky in this PR.

annezazu commented:
>Update: I'll be AFK off and on over the next couple of weeks, so just jotting down some notes and thoughts for next things to work on with this PR, for when I come back to it

Just want to check in as I'm doing the weekly update for phase 2 tasks -- do you need any extra help or back up considering the on and off AFK? Would love to make momentum continues.

andrewserong commented:
> do you need any extra help or back up considering the on and off AFK?

Thanks for checking in @annezazu — Joen and James have added useful feedback for when I’m back, so I plan to dig into those ideas. There are a few things to figure out with this proposed feature, as I imagine we’re going to encounter quite a few edge cases!

The main thing that I think would help with momentum would be for folks to test out this PR, have a go at setting some Group blocks to “sticky” and report back usability thoughts and feedback, that’ll help give a fuller picture of the edge cases that’ll need to be addressed. A good use case for folks to test is “I want a site with a sticky header, can I achieve this by setting a Group block in the header area in the site editor to use sticky + top position”. If the answer is “no”, then let’s figure out the missing things required to getting it working acceptably.

In terms of help with the code side of things, if anyone is free to jump in and wants to work on the code — feel free to fork this PR or commit directly to it, though I understand it can be tricky to jump into a WIP, so I’ve got some other suggestions below. I’ll comment here again when I’m back, so happy to work with anyone who might’ve picked anything up. However, the most useful things to look at I think would be:

* If anyone would like to investigate how to make the block toolbar play nicely with sticky/fixed blocks — when you select a sticky block, the block toolbar should be in an easy to reach place and be visible at all times when scrolling in the block editor while the block is selected. I’m not sure what will be involved in fixing this issue, so that would be the most helpful task from my perspective.
* A dependency for this PR is to switch the LayoutPanel over to use the ToolsPanel so that position controls can be hidden by default. This change will need to land before this PR will become viable. My plan was to pick up this task when I’m back, but I’d be very happy if anyone else wanted to give it a go: https://github.com/WordPress/gutenberg/issues/44560

Hope that helps! Let me know if you need any more info here.

paaljoachim commented:
Testing
Using the Gutenberg PR build.
Twenty Twenty Three
WordPress 6.1
Site Editor

### Before:
![Screenshot 2022-11-08 at 10 45 01](https://user-images.githubusercontent.com/5323259/200532288-c829ba96-2520-4541-8846-45b78595d2f9.jpg)

### After:

Adding sticky to Header.
Clicking top area icon.
Adding a green color to the background.

![Screenshot 2022-11-08 at 10 56 13](https://user-images.githubusercontent.com/5323259/200534102-c801c1b7-3f2d-4652-8560-09f6ae1e3600.jpg)

Sticky did not work in the backend or frontend.

This is what I looked for.
Clicking Position I expected the border around the sticky Group block to change to reflect that I have done something to modify the Group block. Perhaps a dashed border outline to show that the state of the Group block has changed from Normal to Sticky.

jameskoster commented:
> A good use case for folks to test is “I want a site with a sticky header, can I achieve this by setting a Group block in the header area in the site editor to use sticky + top position”

Could we amend that to simply: "I want to make a Group block stick to the top of the viewport as I scroll past it"? Throwing template parts (particularly footers) into the mix at this stage is making things overly complex. Adding the `position: sticky` functionality is a nice enhancement on its own (the Table of Contents block will appreciate!) and we need to get that right before moving on to the more intricate use cases.

Joen updated https://github.com/WordPress/gutenberg/issues/30121 with the latest mockups.

andrewserong commented:
> Could we amend that to simply: "I want to make a Group block stick to the top of the viewport as I scroll past it"? Throwing template parts (particularly footers) into the mix at this stage is making things overly complex.

Great point @jameskoster. Yes, I think for this PR let's limit the scope to the Group block, whatever we can do to implement this incrementally sounds good 👍.

> Joen updated https://github.com/WordPress/gutenberg/issues/30121 with the latest mockups.

Awesome, thanks for that. Just left a follow-up comment over there to make sure I understand the goal for the initial implementation 🙂

> Sticky did not work in the backend or frontend.

Thanks for testing @paaljoachim! I think in that example it's to do with sticky being applied at a non-root level so the area that it affects isn't site wide. I think for the moment this means that (at least with this PR) we're better off pursuing the suggestion James made to just focus on the Group block at the moment, rather than template parts.

Then, once this PR lands we can look at the flow for switching on position for template parts (or a group block that is a parent of a template part) in a follow up.

andrewserong committed:
Update UI controls

andrewserong commented:
Update: I've updated this PR and simplified the UI controls to match the latest designs in #30121 — toggling between Default and Sticky now also sets `top: 0` as well as `position: sticky`, so it seems to be working pretty well now without a separate Area or Placement control: (Note: it does sort of stick out there at the bottom of the list of controls, but it sorta looks a little odd at the top as well, happy for any feedback there)

<img width="1116" alt="image" src="https://user-images.githubusercontent.com/14988353/202981476-a667174e-3f5a-4623-ac2c-2b32e97b103d.png">

It does still take up a fair bit of room for a control that will (most likely) be used only in a couple of places, so I've progressed forward with the idea of converting the LayoutPanel over to using the ToolsPanel in a separate PR over in: https://github.com/WordPress/gutenberg/pull/45833 — that seems to have been going pretty well, the only major issue so far is that there isn't quite enough room between the Justification and Orientation controls for the gutter. I've added a screenshot in this comment: https://github.com/WordPress/gutenberg/pull/45833#issuecomment-1321496372

As for the remaining issues in this PR, next up I'll take a look at:

* The issue of the block toolbar's positioning when you go to scroll down the page (currently the toolbar scrolls off the screen for sticky blocks)
* Z-index tweaking

ramonjd commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -0,0 +1,201 @@
+
+	if ( position === 'sticky' || position === 'fixed' ) {
+		// TODO: Work out where to put the magic z-index value.
+		output += `z-index: 250`;
```
Could the z-index be defined in theme.json `layout.definitions`, e.g., `is-layout-sticky`  similar to `is-layout-flow`?

🤷

andrewserong replied:
Just jotting down a note here that it might make sense to have some preset-like values for `z-index` stored in `settings.layout.zIndexSticky` or if we want support for multiple z-indexes, then perhaps an array of z-indexes for different purposes. That way we could have a core default z-index value for sticky / fixed position, but themes could then override it with their own values if need be.

jameskoster commented:
Thanks for updating this, it seems to be working well. 

We probably need to do something with the helper text – at the moment it's only describing the sticky option. For now it may even make sense to use a switch here:

<img width="299" alt="Screenshot 2022-11-21 at 11 01 07" src="https://user-images.githubusercontent.com/846565/203034461-839388f2-f0fc-4c14-8851-415da188d1f0.png">

I think this may be more intuitive. I acknowledge that it prohibits fixed positions, but I don't think that's a big problem for now, as that use case needs further design consideration. What do you think @jasmussen ?

jasmussen commented:
Thanks for the PR, thanks for the ping, taking for a spin and seeing the toggle work really well:

![status](https://user-images.githubusercontent.com/1204802/203052395-9a427fab-65ca-47bf-811e-730329b8d7cb.gif)

This is super nice. 

* I think Jay may be right, that a simple toggle may be better fow now. It has a smaller visual footprint.
* Unless you're scrolled, there's no visual effect in the canvas. That suggests to me we should a) have it be a non default toolspanel control (✅ I've seen the separate PR) and that we do need the help text.

But perhaps we don't need the "Position" subheading. Just like how there isn't a "content width" subheading for this:

<img width="263" alt="Screenshot 2022-11-21 at 13 29 54" src="https://user-images.githubusercontent.com/1204802/203055951-a7b347b5-ca7b-49d3-a1da-41c02dd27bd0.png">

— just the toggle, and maybe it could say "Stick to the top"?

jasmussen commented:
So, thinking more about this, and from good advice from a friend, it behooves us to consider an interface that can at least scale to additional properties in the future, and perhaps it would be frustrating to see the toggle replaced with a segmented control at a later time. So I tried one more option:

<img width="1928" alt="Sticky toggle" src="https://user-images.githubusercontent.com/1204802/203090963-d773354a-2389-4dc2-914f-6a020777fe09.png">

* It keeps the toggle, with two options, Sticky and Fixed.
* "Default" is gone, because it would be the same as removing the control from the ToolsPanel.
* A small control grid appears on "Fixed", similar to that which exists in the Cover block toolbar, which allows a basic 9 grid of positions for the fixed position control.

As far as near term, we could still go with just "Sticky" as a starting point, and have "Default" and "Sticky" options inside the segmented control. Or we could simplify the positional control with something like this:

<img width="1004" alt="fixed toggle" src="https://user-images.githubusercontent.com/1204802/203092215-f03bdafa-45a4-45e0-92da-f378ff83c267.png">

But it would still be useful to consider a design that can naturally evolve to the additional controls. Let me know your thoughts! @jameskoster I'd love your thoughts on this. It breaks a few rules 😅 [Figma](https://www.figma.com/file/vmHxwa8WhiyiP9poe2om9J/Inspector-Panels?node-id=13565%3A53143&t=TkkUd67zXkUIXhxa-1)

andrewserong commented:
Thanks for the extra discussion! Happy to implement whichever of the designs folks think will work for the first iteration of the feature.

I think if we use a single on/off toggle, then I agree that removing the Position label sounds good. If we retain the `Default/Sticky` options, then keeping the heading seems better to me. I also like the idea of retaining `Default` in the UI in some way, even if we add Fixed as an extra option in a follow-up — as a user, I'd like to be able to toggle it off immediately without having to dig into the ToolsPanel menu.

I like the idea of the X and Y position offsets, too — that'll allow some pretty cool advanced behaviour further down the track. To begin with, though, I reckon if we can keep it to "switch to sticky sets it to `top: 0`" that'll help us with keeping things simple. I really love seeing these explorations for where we might take things in follow-ups, though! 😀

jameskoster commented:
@jasmussen It all depends on how far we want to take this, which isn't 100% clear to me. Do we want a control(s) that is able to entertain the full gamut of css `position` (including `absolute` and `relative` which both have their place)? Or do we want UIs that are tailored for specific use cases (IE fixed header / footer)? Or do we want both? :) 

I'm happy to work on whatever, but it's quite a big task and will be tricky to get right. That's kind of why I suggested the simple toggle to begin with, but maybe it's not useful enough on its own?

jasmussen commented:
> Do we want a control(s) that is able to entertain the full gamut of css position (including absolute and relative which both have their place)? Or do we want UIs that are tailored for specific use cases (IE fixed header / footer)? Or do we want both? :)

I think the answer is that it isn't _entirely_ clear how far we'd want to go, for example my current instinct is that we should avoid "relative" as an option, and ideally provide an interface that makes you not need to worry about relative at all. 

I was also reminded that in one design we have items in all 4 corners, like this:

![8Gg-WI59th](https://user-images.githubusercontent.com/1204802/203312084-ae1d907e-a6de-4674-98ae-b65759eb3d09.png)

How's that accomplished? Is it a two full-width flex-between navigation bars, top and bottom fixed respectively? Or is it 4, each of which is fixed to a corner? And are those positioned using a simple cover-block like matrix, or are those easier to accomplish with x/y coordinates? Or even top/right/bottom/left inputs?

We don't have to accomplish all of this, I still think it's a great idea to start with as little as possible, e.g. sticky as it doesn't require any positioning. But what I realized is that it would probably be a better interface if what we ship today can naturally evolve into what we might ship next year. I.e. if we know we'll need a segmented control in the future, perhaps we shouldn't start with a toggle — even if initially that seemed like a simpler design. What do you think?

jameskoster commented:
> if we know we'll need a segmented control in the future

My point is simply that I'm not convinced we _do_ know that. But if your conviction is strong then let's stick with it because I haven't explored this issue with a microscope just yet 😁 

Otherwise we have two options:

1. Go with the toggle for now, and revise it later
2. Postpone implementation while we perform a deeper dive on the design side

I have no strong feelings either way.

jasmussen commented:
> My point is simply that I'm not convinced we do know that. But if your conviction is strong then let's stick with it because I haven't explored this issue with a microscope just yet 😁

That's the thing, I don't know for sure! But it's the best of the options I [came up with](https://www.figma.com/file/vmHxwa8WhiyiP9poe2om9J/Inspector-Panels?node-id=13565%3A53143&t=mNiPbfDxphGYaIJb-1), mainly because it's hard to find icons to denote the nuance between "fixed" vs. "sticky". A dropdown as also explored previously is another option:

<img width="412" alt="Screenshot 2022-11-22 at 14 12 47" src="https://user-images.githubusercontent.com/1204802/203322671-3577ab1d-8cc8-4140-955c-d5ac2c0bbfc5.png">

Mainly my point is, I would think we have reasonable confidence that we do want to add _fixed_ (plus some positioning controls) in the future, and in that case, the toggle wouldn't scale.

jameskoster commented:
Okay. The only other problem I see is that the segmented control needs to include at least 2 values, so we'll need to either:

1. Implement sticky and fixed positioning at the same time.
2. Include a 'Default' or 'Static' value.

Did you have a preference?

jasmussen commented:
Right, that's the challenge of the moment, where "Default" really doesn't make too much sense if the control is a non-default control. But I think we can probably include it regardless, even without Fixed, — just like how you can add the Dropcap control and toggle it off. Then when we get a chance to add Fixed in the future, we can choose whether we want 3 options, or can do without the Default at that point. What do you think?

jameskoster commented:
So you mean this?

<img width="805" alt="Screenshot 2022-11-22 at 14 37 34" src="https://user-images.githubusercontent.com/846565/203341653-f035eb65-bbb3-48f4-aefe-ce5ce728c51e.png">

I guess it's fine, the only part I'm not sure about is the help text. In this context shouldn't the description speak about the control itself, rather than the specific values therein?

andrewserong commented:
Great questions and discussion, folks! Just my 2 cents, but since it's very easy for controls to become quite complex as we consider more advanced use cases, I think I'd lean toward a simple control to begin with (focusing on the main use case of stick-to-top), accepting that we'll likely wind up exploring a very different control at some point next year when we attempt the more complex use cases.

> In this context shouldn't the description speak about the control itself, rather than the specific values therein?

Good question — I suppose we can have the help text be descriptive of the current state, and have it say something different depending on the option currently selected, a bit like we do for the help text next to the "Inner blocks use content width" control?

jameskoster commented:
I opened a discussion [here](https://github.com/WordPress/gutenberg/discussions/46032) to further discuss the design.

andrewserong commented:
Thanks for all the feedback on this PR, folks! Based on the latest discussions, and since the technical implementation needed to change a bit to decouple from the Layout block support, I've opened up a separate PR for continued exploration over in: https://github.com/WordPress/gutenberg/pull/46142 — that PR is now the current one for future work on this feature (with discussion happening over in https://github.com/WordPress/gutenberg/discussions/46032), so I'll close out this PR now.
